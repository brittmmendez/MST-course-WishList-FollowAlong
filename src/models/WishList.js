import { types, getParent, destroy } from "mobx-state-tree"

export const WishListItem = types
    .model({
        name: types.string,
        price: types.number,
        image: ""
    })
    .actions(self => ({
        changeName(newName) {
            self.name = newName
        },
        changePrice(newPrice) {
            self.price = newPrice
        },
        changeImage(newImage) {
            self.image = newImage
        },
        remove() {
            getParent(self, 2).remove(self)
        }
    }))

export const WishList = types
    .model({
        items: types.optional(types.array(WishListItem), [])
    })
    .actions(self => ({
        add(item) {
            self.items.push(item)
        },
        remove(item) {
            destroy(item)
        }
    }))
    .views(self => ({
        get subTotalPrice() {
            return self.items.reduce((sum, entry) => sum + entry.price, 0)
        },

        get taxes() {
            return parseFloat(self.subTotalPrice * 0.07).toFixed(2)
        },
        get totalPrice() {
            return self.items.reduce((sum, entry) => sum + entry.price, 0)
        }
    }))
