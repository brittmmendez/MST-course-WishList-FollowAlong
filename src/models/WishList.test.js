import { getSnapshot, onSnapshot, onPatch} from "mobx-state-tree"
import { reaction } from "mobx"
import { WishListItem, WishList } from "./WishList"

// WISHILIST ITEM
const item = WishListItem.create({
    "name" : "Chron of Narnia Box Set",
    "price": 28.75,
    "image": "https://www.google.com/search?q=flower+pictures&rlz=1C5CHFA_enUS804US804&tbm=isch&source=iu&ictx=1&fir=yd62M53yvhtdBM%253A%252C5UbOpOqf9qM23M%252C_&usg=AFrqEzc3xs2BwNc35psQ7ZIL4FGZeXx0LQ&sa=X&ved=2ahUKEwi3i6HGjODcAhWo6IMKHe9xBjIQ9QEwAHoECAQQBA#imgrc=yd62M53yvhtdBM:"
})

it("can create an instance of a WishListItem model with correct name", () =>{
    expect(item.name).toBe("Chron of Narnia Box Set")   
})

it("can create an instance of a WishListItem model with correct price", () =>{
    expect(item.price).toBe(28.75)
})

it("can create an instance of a  model with correct image", () =>{
    expect(item.image).toBe("https://www.google.com/search?q=flower+pictures&rlz=1C5CHFA_enUS804US804&tbm=isch&source=iu&ictx=1&fir=yd62M53yvhtdBM%253A%252C5UbOpOqf9qM23M%252C_&usg=AFrqEzc3xs2BwNc35psQ7ZIL4FGZeXx0LQ&sa=X&ved=2ahUKEwi3i6HGjODcAhWo6IMKHe9xBjIQ9QEwAHoECAQQBA#imgrc=yd62M53yvhtdBM:")
})

it("can change the name of an item instance", () =>{
    item.changeName("This is my new name!")
    expect(item.name).toBe("This is my new name!")
})

it("can change the price of an item instance", () =>{
    item.changePrice(22.22)
    expect(item.price).toBe(22.22)
})


// WISHILIST
it("can create a WishList with items", () =>{
    const wishList = WishList.create({
        items: [
            {
                name : "Chron of Narnia Box Set",
                price: 28.75,
                image: "https://www.google.com/search?q=flower+pictures&rlz=1C5CHFA_enUS804US804&tbm=isch&source=iu&ictx=1&fir=yd62M53yvhtdBM%253A%252C5UbOpOqf9qM23M%252C_&usg=AFrqEzc3xs2BwNc35psQ7ZIL4FGZeXx0LQ&sa=X&ved=2ahUKEwi3i6HGjODcAhWo6IMKHe9xBjIQ9QEwAHoECAQQBA#imgrc=yd62M53yvhtdBM:"
            }
        ]
    })
    expect(wishList.items.length).toBe(1)
    expect(wishList.items[0].price).toBe(28.75)
})

it("can create a WishList without any items", () =>{
    const list = WishList.create()
    expect(list.items.length).toBe(0)
})

it("can add an item to a WishList", () =>{
    const emptywishList = WishList.create()
    const states = []
    onSnapshot(emptywishList, snapshot => {
        states.push(snapshot)
    })

    emptywishList.addItem({
            name : "Twilight Box Set",
            price: 28.75
        })
    expect(emptywishList.items.length).toBe(1)
    expect(emptywishList.items[0].price).toBe(28.75)
    
    emptywishList.items[0].changeName("It's Changed!")
    expect(emptywishList.items[0].name).toBe("It's Changed!")

    expect(getSnapshot(emptywishList)).toEqual({
        items: [
            {
                name: "It's Changed!",
                price: 28.75,
                image: ""
            }
        ]
    })

    expect(getSnapshot(emptywishList)).toMatchSnapshot()

    expect(states).toMatchSnapshot()
})

it("can add an item to a WishList -2", () =>{
    const emptywishList = WishList.create()
    const patches = []
    onPatch(emptywishList, patch => {
        patches.push(patch)
    })

    emptywishList.addItem({
            name : "Twilight Box Set",
            price: 28.75
        })
    
    emptywishList.items[0].changeName("It's Changed!")

    expect(patches).toMatchSnapshot()
})


it("can calculate the total price of a wishlist", () =>{
    const list = WishList.create()
    list.addItem({name : "item 1", price: 2})
    list.addItem({name : "item 2", price: 4})
    list.addItem({name : "item 3", price: 5})

    expect(list.totalPrice).toBe(11)

    let changed = 0 
    reaction(() => list.totalPrice, () => changed++)
    expect(changed).toBe(0)
    console.log("total Price" + list.totalPrice)
    console.log("changed:"+ changed)
    list.items[0].changeName("New Item 1")
    expect(changed).toBe(0)
    console.log("total Price" + list.totalPrice)
    console.log("changed:"+ changed)
    list.addItem({name : "item 4", price: 6})
    expect(changed).toBe(1)
    console.log("total Price" + list.totalPrice)
    console.log("changed:"+ changed)

})