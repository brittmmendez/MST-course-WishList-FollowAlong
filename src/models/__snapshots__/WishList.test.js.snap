// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add an item to a WishList -2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Twilight Box Set",
      "price": 28.75,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "It's Changed!",
  },
]
`;

exports[`can add an item to a WishList 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "It's Changed!",
      "price": 28.75,
    },
  ],
}
`;

exports[`can add an item to a WishList 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Twilight Box Set",
        "price": 28.75,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "It's Changed!",
        "price": 28.75,
      },
    ],
  },
]
`;
